adapter/avro/src/main/java/org/apache/arrow/adapter/avro/ArrowToAvroUtils.java

```diff
@@ -0,0 +1,729 @@
+import org.apache.arrow.vector.UInt8Vector;
+import org.apache.arrow.vector.VarBinaryVector;
+import org.apache.arrow.vector.VarCharVector;
+import org.apache.arrow.vector.complex.DenseUnionVector;
+import org.apache.arrow.vector.complex.FixedSizeListVector;
+import org.apache.arrow.vector.complex.ListVector;
+import org.apache.arrow.vector.complex.MapVector;
+import org.apache.arrow.vector.complex.StructVector;
+import org.apache.arrow.vector.complex.UnionVector;
+import org.apache.arrow.vector.types.FloatingPointPrecision;
+import org.apache.arrow.vector.types.TimeUnit;
+import org.apache.arrow.vector.types.Types;
+import org.apache.arrow.vector.types.pojo.ArrowType;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.avro.Schema;
+import org.apache.avro.SchemaBuilder;
+
+public class ArrowToAvroUtils {
+
+  public static final String GENERIC_RECORD_TYPE_NAME = "GenericRecord";
+
+  /**
+   * Create an Avro record schema for a given list of Arrow fields.
+   *
+   * <p>This method currently performs following type mapping for Avro data types to corresponding
+   * Arrow data types.
+   *
+   * <ul>
+   *   <li>ArrowType.Null --> NULL
+   *   <li>ArrowType.Bool --> BOOLEAN
+   *   <li>ArrowType.Int(64 bit, unsigned 32 bit) --> LONG
+   *   <li>ArrowType.Int(signed 32 bit, < 32 bit) --> INT
+   *   <li>ArrowType.FloatingPoint(double) --> DOUBLE
+   *   <li>ArrowType.FloatingPoint(single, half) --> FLOAT
+   *   <li>ArrowType.Utf8 --> STRING
+   *   <li>ArrowType.LargeUtf8 --> STRING
+   *   <li>ArrowType.Binary --> BYTES
+   *   <li>ArrowType.LargeBinary --> BYTES
+   *   <li>ArrowType.FixedSizeBinary --> FIXED
+   *   <li>ArrowType.Decimal --> decimal (FIXED)
+   *   <li>ArrowType.Date --> date (INT)
+   *   <li>ArrowType.Time (MILLI) --> time-millis (INT)
+   *   <li>ArrowType.Time (SEC | MICRO | NANO) --> time-micros (LONG)
+   *   <li>ArrowType.Timestamp (NANOSECONDS, TZ != NULL) --> time-nanos (LONG)
+   *   <li>ArrowType.Timestamp (MICROSECONDS, TZ != NULL) --> time-micros (LONG)
+   *   <li>ArrowType.Timestamp (MILLISECONDS | SECONDS, TZ != NULL) --> time-millis (LONG)
+   *   <li>ArrowType.Timestamp (NANOSECONDS, TZ == NULL) --> local-time-nanos (LONG)
+   *   <li>ArrowType.Timestamp (MICROSECONDS, TZ == NULL) --> local-time-micros (LONG)
+   *   <li>ArrowType.Timestamp (MILLISECONDS | SECONDS, TZ == NULL) --> local-time-millis (LONG)
+   *   <li>ArrowType.Duration --> duration (FIXED)
+   *   <li>ArrowType.Interval --> duration (FIXED)
+   *   <li>ArrowType.Struct --> record
+   *   <li>ArrowType.List --> array
+   *   <li>ArrowType.LargeList --> array
+   *   <li>ArrowType.FixedSizeList --> array
+   *   <li>ArrowType.Map --> map
+   *   <li>ArrowType.Union --> union
+   * </ul>
+   *
+   * <p>Nullable fields are represented as a union of [null | base-type]. Special treatment is given
+   * to nullability of unions - a union is considered nullable if the union field is nullable or any
```

Comment:
FWIW, I think nullable unions aren't semantically valid - unions have no validity bitmap in Arrow. The only way to represent this is with a nullable child field

Reply:
I checked the spec and yes you are correct!  I have updated the logic accordingly

Reply:
Can we update the docstring too?

Reply:
I have updated the comment on unions. Also updated the comment for nullable types, which are encoded as [ base-type | null ], per the discussion above.